vg/vgtex/canvas.go

```diff
@@ -0,0 +1,291 @@
+	c.ctx().dashArray = pattern
+	c.ctx().dashOffset = offset
+}
+
+// SetColor implements the vg.Canvas.SetColor method.
+func (c *Canvas) SetColor(clr color.Color) {
+	c.ctx().color = clr
+}
+
+// Rotate implements the vg.Canvas.Rotate method.
+func (c *Canvas) Rotate(rad float64) {
+	c.wtex(`\pgftransformrotate{%g}`, rad*degPerRadian)
+}
+
+// Translate implements the vg.Canvas.Translate method.
+func (c *Canvas) Translate(x, y vg.Length) {
+	c.wtex(`\pgftransformshift{\pgfpoint{%gpt}{%gpt}}`, x, y)
+}
+
+// Scale implements the vg.Canvas.Scale method.
+func (c *Canvas) Scale(x, y float64) {
+	c.wtex(`\pgftransformxscale{%g}`, x)
+	c.wtex(`\pgftransformyscale{%g}`, y)
+}
+
+// Push implements the vg.Canvas.Push method.
+func (c *Canvas) Push() {
+	c.wtex(`\begin{pgfscope}`)
+	c.stack = append(c.stack, *c.ctx())
+}
+
+// Pop implements the vg.Canvas.Pop method.
+func (c *Canvas) Pop() {
+	c.stack = c.stack[:len(c.stack)-1]
+	c.wtex(`\end{pgfscope}`)
+	c.wtex("")
+}
+
+// Stroke implements the vg.Canvas.Stroke method.
+func (c *Canvas) Stroke(p vg.Path) {
+	if c.ctx().linew.Dots(c.DPI()) <= 0 {
+		return
+	}
+	c.wstyle()
+	c.wpath(p)
+	c.wtex(`\pgfusepath{stroke}`)
+	c.wtex("")
+}
+
+// Fill implements the vg.Canvas.Fill method.
+func (c *Canvas) Fill(p vg.Path) {
+	c.wstyle()
+	c.wpath(p)
+	c.wtex(`\pgfusepath{fill, stroke}`)
+	c.wtex("")
+}
+
+// FillString implements the vg.Canvas.FillString method.
+func (c *Canvas) FillString(f vg.Font, x, y vg.Length, text string) {
+	c.wcolor()
+	c.wtex(`\pgftext[base,at={\pgfpoint{%gpt}{%gpt}}]{%s}`,
+		x+f.Width(text)/2,
+		y,
+		text,
+	)
+}
+
+func (c *Canvas) indent(s string) string {
+	return strings.Repeat(s, len(c.stack))
+}
+
+func (c *Canvas) wtex(s string, args ...interface{}) {
+	fmt.Fprintf(c.buf, c.indent("  ")+s+"\n", args...)
+}
+
+func (c *Canvas) wstyle() {
+	c.wdash()
+	c.wlineWidth()
+	c.wcolor()
+}
+
+func (c *Canvas) wdash() {
+	if len(c.ctx().dashArray) == 0 {
+		return
+	}
+	str := `\pgfsetdash{`
+	for _, d := range c.ctx().dashArray {
+		str += fmt.Sprintf("{%gpt}", d)
+	}
+	str += fmt.Sprintf("}{%gpt}", c.ctx().dashOffset)
+	c.wtex(str)
+}
+
+func (c *Canvas) wlineWidth() {
+	c.wtex(`\pgfsetlinewidth{%gpt}`, c.ctx().linew)
+}
+
+func (c *Canvas) wcolor() {
+	col := c.ctx().color
+	if col == nil {
+		col = color.Black
+	}
+	r, g, b, a := col.RGBA()
+	alpha := 255.0 / float64(a)
+	// FIXME(sbinet) \color will last until the end of the current TeX group
+	// use \pgfsetcolor and \pgfsetstrokecolor instead.
+	// it needs a named color: define it on the fly (storing it at the beginning
+	// of the document.)
+	c.wtex(
+		`\color[rgb]{%g,%g,%g}`,
+		float64(r)*alpha/255.0,
+		float64(g)*alpha/255.0,
+		float64(b)*alpha/255.0,
+	)
+
+	opacity := float64(a) / math.MaxUint16
+	c.wtex(`\pgfsetstrokeopacity{%g}`, opacity)
+	c.wtex(`\pgfsetfillopacity{%g}`, opacity)
+}
+
+func (c *Canvas) wpath(p vg.Path) {
+	for _, comp := range p {
+		switch comp.Type {
+		case vg.MoveComp:
+			c.wtex(
```

Comment:
What is the style we're going with in this file? At the moment it's not consistent. Is there any real reason not to have all the c.wtex calls with a single line?